src/anisotropic_smoothing.cpp

```diff
@@ -0,0 +1,79 @@
+#include <arrayfire_benchmark.h>
+#include <benchmark/benchmark.h>
+#include <arrayfire.h>
+
+#include <cmath>
+#include <vector>
+
+using std::vector;
+using af::array;
+using af::randu;
+using af::anisotropicDiffusion;
+using af::dim4;
+using af::deviceMemInfo;
+using af::deviceGC;
+
+static
+void asmBase(benchmark::State& state,
+             dim4 dims,
+             af_dtype type,
+             float timestep,
+             float conductance,
+             unsigned iterations)
+{
+    array inImg = randu(dims, type);
+
+   //allocate output once to bypass alloc calls
+   //when smoothing function is actually called
+   if (type==f64)
+       array outTemp = randu(dims, f64);
+   else
+       array outTemp = randu(dims, f32);
+
+  size_t alloc_bytes, alloc_buffers, lock_bytes, lock_buffers;
+  deviceMemInfo(&alloc_bytes, &alloc_buffers, &lock_bytes, &lock_buffers);
+
+  for (auto _ : state) {
+      array out = anisotropicDiffusion(inImg, timestep, conductance, iterations);
+      af::sync();
+  }
+
+  size_t alloc_bytes2, alloc_buffers2, lock_bytes2, lock_buffers2;
+  deviceMemInfo(&alloc_bytes2, &alloc_buffers2, &lock_bytes2, &lock_buffers2);
+
+  state.counters["alloc_bytes"] = alloc_bytes2 - alloc_bytes;
+  state.counters["alloc_buffers"] = alloc_buffers2 - alloc_buffers;
+  deviceGC();
+}
+
+static
+void asmBench(benchmark::State& state, af_dtype type)
+{
+    unsigned dim   = state.range(0);
+    unsigned iters = state.range(1);
+    asmBase(state, af::dim4(dim, dim), type, 0.0105833, 0.35, iters);
+}
+
+int main(int argc, char** argv)
+{
+    vector<af_dtype> types = {f32, u32, u16, u8};
+
+    af::benchmark::RegisterBenchmark("asm_ImageSizes", types, asmBench)
+        ->RangeMultiplier(2)
+        ->Ranges({{16, 8192}, {32, 32}})
+        ->ArgNames({"dim", "asm_iterations"})
+        ->Iterations(100)
```

Comment:
Not setting the iteration count is basically letting the first run time skew the other results. We have to run it for many iterations to average that out like `timeit` and otherwise, run some warp up code.

Reply:
Also, if iterations are left to the benchmark code, it is being set to as high as 20K+ some times which I don't is necessary in any circumstance for our code.

Reply:
Leaving it black will allow the ability to handle different cards. For example if you run the same code on the V100 vs GT750 or targeting the CPU backend. This approach will significantly take a long time on smaller systems.

Reply:
True, I didn't account for CPU backend. But leaving it blank is equally worse some times whether it is GT series card or Tesla series card because it is taking too much time to execute the benchmarks as the iteration count is as high as 20k+ especially when a single run time is very small. This will quickly add up to a lot when we run all benchmarks. I am removing it for the moment, but we should really think about this instead of leaving it out for gbench to figure out.